lib/luminork-server/src/search/query.rs

```diff
@@ -0,0 +1,317 @@
+use std::str::FromStr;
+
+use ulid::Ulid;
+
+use crate::search::{
+    Error,
+    Result,
+    parser,
+};
+
+/// A search query
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum SearchQuery {
+    MatchValue(SearchTerm),
+    MatchAttr {
+        name: String,
+        terms: Vec<SearchTerm>,
+    },
+    And(Vec<SearchQuery>),
+    /// One of the sub-queries must match
+    Or(Vec<SearchQuery>),
+    /// The sub-query must not match
+    Not(Box<SearchQuery>),
+    /// Used for the empty query. Matches everything.
+    All,
+}
+
+/// A value to match
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum SearchTerm {
+    /// Match the string, allowing * and such: Instance, AWS::*::*Group
+    Match(String),
+    /// Exact literal match: "AWS::EC2::Instance"
+    Exact(String),
+    /// Match starting with the given string: "AWS::EC2::Inst
+    StartsWith(String),
+}
+
+impl FromStr for SearchQuery {
+    type Err = Error;
+
+    fn from_str(s: &str) -> Result<Self> {
+        parser::parse(s)
+    }
+}
+
+impl SearchTerm {
```

Comment:
These utilities are designed to handle case sensitivity, number/boolean matching, etc. in a consistent way. That way we can test this stuff in one place, and the actual component-specific code is smaller and easier to test in isolation. This also will let us reuse a lot of this logic when we implement search for other things (schemas, functions, etc.)